Change default keysets to be static and atomic

sypneiwski opened the pull request:
This changes the default keysets to be mutable and global, as discussed. I'm using an atomic instead of locks for performance. 

The operations on the keysets should be mostly thread-safe, however it is possible in theory that a thread can modify the keyset after another thread loads but before it stores its keyset, resulting in an overwrite. I'm relying on the keysets not being modified often and the callers synchronizing with each other. If needed, I can look into improving the safety.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85543
* #85469
* #85468
* __->__ #85160

sypneiwski committed:
Change default keysets to be static and atomic

[ghstack-poisoned]

sypneiwski committed:
Update on "Change default keysets to be static and atomic"

This changes the default keysets to be mutable and global, as discussed. I'm using an atomic instead of locks for performance. 

The operations on the keysets should be mostly thread-safe, however it is possible in theory that a thread can modify the keyset after another thread loads but before it stores its keyset, resulting in an overwrite. I'm relying on the keysets not being modified often and the callers synchronizing with each other. If needed, I can look into improving the safety.




[ghstack-poisoned]

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -37,18 +37,18 @@ struct C10_API PODLocalDispatchKeySet {
   }
   void set_excluded(DispatchKeySet x) {
-    excluded_ = (x ^ c10::default_excluded_set).raw_repr();
+    excluded_ = (x ^ get_default_excluded_set()).raw_repr();
```
This is now a function call; I want this to be only a memory dereference with this change. Also, you don't need a global exclude, so I wouldn't pay the cost for the memory deref for exclude unless you need it

sypneiwski committed:
Update on "Change default keysets to be static and atomic"

This changes the default keysets to be mutable and global, as discussed. I'm using an atomic instead of locks for performance. 

The operations on the keysets should be mostly thread-safe, however it is possible in theory that a thread can modify the keyset after another thread loads but before it stores its keyset, resulting in an overwrite. I'm relying on the keysets not being modified often and the callers synchronizing with each other. If needed, I can look into improving the safety.




[ghstack-poisoned]

ezyang commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -4,6 +4,19 @@
 
 namespace c10 {
 
+// On x86 architecture atomic operations are lock-less.
```
We also build on ARM. But a relaxed concurrency should be acceptable here. More below.

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -37,15 +37,15 @@ struct C10_API PODLocalDispatchKeySet {
   DispatchKeySet included() const {
     return DispatchKeySet(DispatchKeySet::RAW, included_) ^
-        c10::default_included_set;
+        c10::default_included_set.load();
```
I believe `memory_order_relaxed` is acceptable here. The model for when it is acceptable to change the default included set is at the very start of the program, before any threads hitting the dispatcher have begun. In this case, there is an implicit synchronization between the setter and all subsequent uses, and therefore no race is possible and relaxed is OK.

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -37,15 +37,15 @@ struct C10_API PODLocalDispatchKeySet {
 
   void set_included(DispatchKeySet x) {
-    included_ = (x ^ c10::default_included_set).raw_repr();
+    included_ = (x ^ c10::default_included_set.load()).raw_repr();
```
Ditto here

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -645,10 +646,10 @@ constexpr DispatchKeySet autocast_dispatch_keyset = DispatchKeySet({
-    DispatchKey::BackendSelect,
-    DispatchKey::ADInplaceOrView,
-});
+extern C10_API std::atomic<DispatchKeySet> default_included_set;
```
I hope this works on Windows

ezyang commented:
Please inspect the generated assembly and post it here.

sypneiwski committed:
Update on "Change default keysets to be static and atomic"

This changes the default keysets to be mutable and global, as discussed. I'm using an atomic instead of locks for performance. 

The operations on the keysets should be mostly thread-safe, however it is possible in theory that a thread can modify the keyset after another thread loads but before it stores its keyset, resulting in an overwrite. I'm relying on the keysets not being modified often and the callers synchronizing with each other. If needed, I can look into improving the safety.




[ghstack-poisoned]
